tests/TestKdbx3.cpp

```diff
@@ -0,0 +1,209 @@
+#include "format/KdbxXmlWriter.h"
+#include "format/KeePass2Repair.h"
+#include "config-keepassx-tests.h"
+
+#include <QTest>
+
+QTEST_GUILESS_MAIN(TestKdbx3)
+
+void TestKdbx3::initTestCaseImpl()
+{
+
+}
+
+Database* TestKdbx3::readXml(const QString& path, bool strictMode, bool& hasError, QString& errorString)
+{
+    KdbxXmlReader reader(KeePass2::FILE_VERSION_3_1);
+    reader.setStrictMode(strictMode);
+    auto db = reader.readDatabase(path);
+    hasError = reader.hasError();
+    errorString = reader.errorString();
+    return db;
+}
+
+Database* TestKdbx3::readXml(QBuffer* buf, bool strictMode, bool& hasError, QString& errorString)
+{
+    KdbxXmlReader reader(KeePass2::FILE_VERSION_3_1);
+    reader.setStrictMode(strictMode);
+    auto db = reader.readDatabase(buf);
+    hasError = reader.hasError();
+    errorString = reader.errorString();
+    return db;
+}
+
+void TestKdbx3::writeXml(QBuffer* buf, Database* db, bool& hasError, QString& errorString)
+{
+    KdbxXmlWriter writer(KeePass2::FILE_VERSION_3_1);
+    writer.writeDatabase(buf, db);
+    hasError = writer.hasError();
+    errorString = writer.errorString();
+}
+
+void TestKdbx3::readKdbx(QIODevice* device, CompositeKey const& key, QScopedPointer<Database>& db,
+                         bool& hasError, QString& errorString)
+{
+    KeePass2Reader reader;
+    db.reset(reader.readDatabase(device, key));
+    hasError = reader.hasError();
+    if (hasError) {
+        errorString = reader.errorString();
+    }
+    QCOMPARE(reader.version(), KeePass2::FILE_VERSION_3_1 & KeePass2::FILE_VERSION_CRITICAL_MASK);
+}
+
+void TestKdbx3::readKdbx(const QString& path, CompositeKey const& key, QScopedPointer<Database>& db,
+                         bool& hasError, QString& errorString)
+{
+    KeePass2Reader reader;
+    db.reset(reader.readDatabase(path, key));
+    hasError = reader.hasError();
+    if (hasError) {
+        errorString = reader.errorString();
+    }
+    QCOMPARE(reader.version(), KeePass2::FILE_VERSION_3_1 & KeePass2::FILE_VERSION_CRITICAL_MASK);
+}
+
+void TestKdbx3::writeKdbx(QIODevice* device, Database* db, bool& hasError, QString& errorString)
+{
+    KeePass2Writer writer;
+    hasError = writer.writeDatabase(device, db);
+    hasError = writer.hasError();
+    if (hasError) {
+        errorString = writer.errorString();
+    }
+    QCOMPARE(writer.version(), KeePass2::FILE_VERSION_3_1);
+}
+
+void TestKdbx3::testFormat300()
+{
+    QString filename = QString(KEEPASSX_TEST_DATA_DIR).append("/Format300.kdbx");
+    CompositeKey key;
+    key.addKey(PasswordKey("a"));
+    KeePass2Reader reader;
+    QScopedPointer<Database> db(reader.readDatabase(filename, key));
+    QCOMPARE(reader.version(), KeePass2::FILE_VERSION_3);
+    QVERIFY(db.data());
+    QVERIFY(!reader.hasError());
+
+    QCOMPARE(db->rootGroup()->name(), QString("Format300"));
+    QCOMPARE(db->metadata()->name(), QString("Test Database Format 0x00030000"));
+}
+
+void TestKdbx3::testNonAscii()
```

Comment:
@phoerious is it only the password that is non ascii? in which case we should rename that `testNonAsciiPassword`

Reply:
I think it's just the password, but I'm not entirely sure. It's one of the tests I copied over.